web/scalica/micro/models.py

```diff
@@ -1,3 +1,68 @@
+from django.db import models
+from django.contrib.auth.models import User
+
+class Profile(models.Model):
+    user = models.OneToOneField(User, on_delete=models.CASCADE)
+    first_name = models.CharField(max_length=30)
+    last_name = models.CharField(max_length=30)
+    email = models.CharField(max_length=255)
+    school = models.ForeignKey('School', on_delete=models.SET_NULL) # A user can have no school, but then they shouldn't be allowed to answer questions
+
+@receiver(post_save, sender=User)
+def create_user_profile(sender, instance, created, **kwargs):
+    if created:
+        Profile.objects.create(user=instance)
+
+@receiver(post_save, sender=User)
+def save_user_profile(sender, instance, **kwargs):
+    instance.profile.save()
+
+class Question(models.Model):
+    timestamp = models.DateField(auto_now_add=True, editable=False)
+    text = models.TextField(max_length=500)
+    school = models.ForeignKey(School, on_delete=models.CASCADE) # no school -> no question
+    user = models.ForeignKey(User, on_delete=models.CASCADE) # no user -> no question
+    duplicate_of = models.ForeignKey('self', on_delete=models.SET_NULL)
+
+    def get_answers_by_upvotes(self):
+        return self.answer_set.order_by('-num_upvotes')
+
+    def get_answers_by_date(self):
+        return self.answer_set.order_by('-timestamp')
+
+class Answer(models.Model):
+    timestamp = models.DateField(auto_now_add=True, editable=False)
+    text = models.TextField(max_length=2000)
+    question = models.ForeignKey(Question, on_delete=models.CASCADE)
+    num_upvotes = models.PositiveIntegerField()
```

Comment:
since we have the `Upvotes` model, should we get rid of `num_upvotes` and instead have a field that stores all the upvotes the answer has?

`upvotes  = [Array of Upvote objects]`

Reply:
I believe arrays aren't allowed in databases? I could be wrong though.

However, because Upvote already has a User foreign key, we don't need any extra fields like num_upvotes or upvotes on the User model, because we can query the DB for any information we need.

For example, if we were to do a query using SQL (pseudocode):
SELECT * FROM Upvote WHERE User = Alice
will return an array of Alice's upvotes. We can also call len() on that to get the number of upvotes Alice has done.

Another more practical example:
SELECT * FROM Upvote WHERE Answer = 1
will return an array of Upvote objects for the Answer with ID 1. We can call len() on that and send the number of upvotes to be rendered on the front end for the answer page.

Reply:
I added it to save from having to make the query since it is somewhat expensive. This is addressed in comment "updating num_upvotes on every Upvote save prevents querying for upvote count (which would be costly when used to filter every answer for a question)".  Seemed like the better option, though I agree it's redundant.

Reply:
I'm good with either way haha. We can ask the professor later.

Reply:
True, I was curious about that so I'll ask him, maybe at the hackathon.